File: frontend/src/app/services/socket.service.ts
import { Injectable, OnDestroy } from '@angular/core';
import { Observable, Subject } from 'rxjs';
import { io, Socket } from 'socket.io-client';
import { environment } from '../../environments/environment';
import { AuthService } from './auth.service';

@Injectable({ providedIn: 'root' })
export class SocketService implements OnDestroy {
  private socket!: Socket;
  private destroy$ = new Subject<void>();

  constructor(private auth: AuthService) {}

  connect(): void {
    if (this.socket?.connected) return;
    this.socket = io(environment.socketUrl, {
      auth: { token: this.auth.token },
      transports: ['websocket'],
      reconnectionAttempts: 5,
      reconnectionDelay: 1000,
    });

    this.socket.on('connect', () => console.log('[Socket] Connected', this.socket.id));
    this.socket.on('disconnect', () => console.log('[Socket] Disconnected'));
    this.socket.on('connect_error', (err) => console.error('[Socket] Error:', err.message));
  }

  disconnect(): void {
    this.socket?.disconnect();
  }

  emit(event: string, data?: any): void {
    this.socket?.emit(event, data);
  }

  on<T = any>(event: string): Observable<T> {
    return new Observable<T>(observer => {
      this.socket?.on(event, (data: T) => observer.next(data));
      return () => this.socket?.off(event);
    });
  }

  get connected(): boolean {
    return this.socket?.connected ?? false;
  }

  ngOnDestroy(): void {
    this.disconnect();
    this.destroy$.next();
    this.destroy$.complete();
  }
}

File: queryquest_app/lib/services/socket_service.dart
import 'dart:async';
import 'package:socket_io_client/socket_io_client.dart' as io;
import 'package:shared_preferences/shared_preferences.dart';
import 'api_service.dart';

class SocketService {
  io.Socket? _socket;
  final StreamController<Map<String, dynamic>> _eventStream = StreamController<Map<String, dynamic>>.broadcast();

  Stream<Map<String, dynamic>> get stream => _eventStream.stream;

  Future<void> connect() async {
    if (_socket != null && _socket!.connected) return;

    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('qq_token');

    _socket = io.io(ApiConfig.socketUrl, <String, dynamic>{
      'transports': ['websocket'],
      'autoConnect': false,
      'reconnectionAttempts': 5,
      'reconnectionDelay': 1000,
      'auth': {'token': token},
    });

    _socket!.connect();

    _socket!.onConnect((_) {
      print('[Socket] Connected ${_socket!.id}');
    });

    _socket!.onDisconnect((_) {
      print('[Socket] Disconnected');
    });

    _socket!.onConnectError((err) {
      print('[Socket] Error: $err');
    });

    // Game Events
    _listenToEvent('countdown');
    _listenToEvent('question');
    _listenToEvent('timer_tick');
    _listenToEvent('answer_result');
    _listenToEvent('scores_update');
    _listenToEvent('game_over');
    _listenToEvent('matchmaking_status');
    _listenToEvent('player_joined');
    _listenToEvent('room_updated');
    _listenToEvent('ready_status');
    _listenToEvent('error');
    _listenToEvent('active_players');
  }

  void _listenToEvent(String eventName) {
    _socket!.on(eventName, (data) {
      _eventStream.add({'event': eventName, 'data': data});
    });
  }

  void disconnect() {
    _socket?.disconnect();
    _socket = null;
  }

  void emit(String event, [dynamic data]) {
    if (_socket != null && _socket!.connected) {
      _socket!.emit(event, data);
    }
  }

  bool get connected => _socket?.connected ?? false;
}
